engines/savestate.cpp

```diff
@@ -32,16 +33,23 @@ SaveStateDescriptor::SaveStateDescriptor()
 	_thumbnail(), _saveType(kSaveTypeUndetermined) {
 }
 
-SaveStateDescriptor::SaveStateDescriptor(int s, const Common::U32String &d)
-	: _slot(s), _description(d), _isDeletable(true), _isWriteProtected(false),
-	  _isLocked(false), _saveDate(), _saveTime(), _playTime(), _playTimeMSecs(0),
-	_thumbnail(), _saveType(kSaveTypeUndetermined) {
+SaveStateDescriptor::SaveStateDescriptor(int slot, const Common::U32String &d)
+	: _slot(slot), _description(d), _isLocked(false), _playTimeMSecs(0) {
+	initSaveType();
 }
 
-SaveStateDescriptor::SaveStateDescriptor(int s, const Common::String &d)
-	: _slot(s), _description(Common::U32String(d)), _isDeletable(true), _isWriteProtected(false),
-	_isLocked(false), _saveDate(), _saveTime(), _playTime(), _playTimeMSecs(0),
-	_thumbnail(), _saveType(kSaveTypeUndetermined) {
+SaveStateDescriptor::SaveStateDescriptor(int slot, const Common::String &d)
+	: _slot(slot), _description(Common::U32String(d)), _isLocked(false), _playTimeMSecs(0) {
+	initSaveType();
+}
+
+void SaveStateDescriptor::initSaveType()
+{
+	// Do not allow auto-save slot to be deleted or overwritten.
+	const bool autosave = g_engine ? _slot == g_engine->getAutosaveSlot() : false;
```

Comment:
This must take autosave_period configuration into account like it's done in MetaEngine::listSaves

Reply:
I thought about it, but I think it's better not to. Consider the following use-case:

* Autosave is disabled - slot 0 is a regular slot.
* The user happily saves on this slot.
* After several weeks, the user discovers the autosave feature and enabled it.
* 5 minutes into the game and Poof - the saved game is gone!

Reply:
In this case it would be best to warn the user when they enable this option, something like: "The autosave uses slot XX, it will be lost. Proceed?"

Reply:
Good point. So when enabling this option, I'll iterate over all the engines and check if they have a saved game which is not named "Autosave" on the autosave slot, and warn on this case?

Reply:
Exactly, something like this.

Reply:
I have another idea, which I'm working on.

When autosave is triggered, I check if this slot is occupied by a non-autosave entry. If it is, I prompt the user to choose between:
* Move the currently saved game to a free slot
* Overwrite
* Disable autosave (for this session)

A similar approach can also be useful to prevent accidental overwrite of autosave. For example, if the existing autosave was saved after 3 hours of gameplay, then the user started a new game and autosave was triggered, we can warn before overwriting the existing one.

What do you say?

Reply:
Done